### PR TITLE
scan added

### PR DESCRIPTION
**Pull request by vulkomilev**
scan functionality added

### CODE REVIEW

File: keras/backend/tensorflow/core.py [hunk 1]
```diff
@@ -194,6 +194,19 @@ def slice_update(inputs, start_indices, updates):
     return dynamic_update_slice(inputs, updates, start_indices)
 
 
+def scan(f,
+         init,
+         xs,
+         length=None,
+         reverse=False,
+         unroll=1):
+    return tf.scan(
```
Comment (fchollet) [inline keras/backend/tensorflow/core.py:203]: Please format the code via `sh shell/format.sh`

File: keras/ops/core.py [hunk 2]
```diff
@@ -204,6 +204,16 @@ def slice_update(inputs, start_indices, updates):
     return backend.core.slice_update(inputs, start_indices, updates)
 
 
+@keras_export("keras.ops.scan")
```
Comment (fchollet) [inline keras/ops/core.py:207]: There should be a correspond symbolic op class.

Also, please add a docstring for this function.

File: keras/ops/core_test.py [hunk 3]
```diff
@@ -215,6 +215,17 @@ def test_slice_update(self):
         outputs = core.slice_update(inputs, start_indices, updates)
         self.assertAllClose(outputs[1:3, 1:3, 2:4, 2:4], np.zeros([2, 2, 2, 2]))
 
+
+    def test_scan(self):
```
Comment (fchollet) [inline keras/ops/core_test.py:219]: Let's use `jax.scan` as the reference and let's test consistency with the reference across multiple inputs and dtypes. Let's also test static shape inference.

**Review by fchollet — COMMENTED** [review]
Thank you for the PR! 👍

File: keras/backend/tensorflow/core.py (hunk 1 above)
Comment (vulkomilev) [reply keras/backend/tensorflow/core.py]: done

File: keras/ops/core.py (hunk 2 above)
Comment (vulkomilev) [reply keras/ops/core.py:215]: done.I am not sure about the corresponding symbolic class

File: keras/ops/core_test.py (hunk 3 above)
Comment (vulkomilev) [reply keras/ops/core_test.py:219]: I have added two tests one with int and one with float. The float is not passing due to rounding errors. Should I round it to remove such errors . Also I am not sure about what do you mean by static shape inference sorry.

File: keras/backend/tensorflow/core.py [hunk 4]
```diff
@@ -194,6 +194,25 @@ def slice_update(inputs, start_indices, updates):
...
 
 
+def scan(f, init, xs, length=None, reverse=False, unroll=1):
+
+    if xs is None:
+        xs = [None] * length
+    if reverse:
+        np.flip(xs)
+
+    init = (init, np.array(0, dtype=init.dtype))
```
Comment (fchollet) [inline keras/backend/tensorflow/core.py:204]: You can use `tf.zeros_like` instead of `np.array`

File: keras/backend/tensorflow/core.py [hunk 5]
```diff
@@ -194,6 +194,25 @@ def slice_update(inputs, start_indices, updates):
...
+
+    carry, ys = tf.scan(f, xs, initializer=init)
+
+    ys = ys.numpy()
```
Comment (fchollet) [inline keras/backend/tensorflow/core.py:208]: Don't convert to numpy, keep TF tensors throughout.

File: keras/backend/tensorflow/core.py [hunk 6]
```diff
@@ -194,6 +194,25 @@ def slice_update(inputs, start_indices, updates):
     return dynamic_update_slice(inputs, updates, start_indices)
 
 
+def scan(f, init, xs, length=None, reverse=False, unroll=1):
+
+    if xs is None:
+        xs = [None] * length
+    if reverse:
+        np.flip(xs)
```
Comment (fchollet) [inline keras/backend/tensorflow/core.py:202]: Use a TF op for this

File: keras/backend/torch/core.py [hunk 7]
```diff
@@ -335,6 +335,21 @@ def slice_update(inputs, start_indices, updates):
...
+    carry = init
+    ys = []
+    if reverse:
+        xs.reverse()
+    for x in xs:
+        carry, y = f(carry, x)
+        ys.append(y)
+    if reverse:
+        ys.reverse()
+    return carry, np.stack(ys)
```
Comment (fchollet) [inline keras/backend/torch/core.py:350]: Use torch ops, not np

File: keras/ops/core_test.py (hunk 3 above)
Comment (fchollet) [reply keras/ops/core_test.py:219]: > Also I am not sure about what do you mean by static shape inference sorry.

This means testing the output shape and dtypes obtained when calling the op on a KerasTensor. There is a test case for this in this file.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

**Comment by vulkomilev** [comment]
@fchollet I have made some changes

File: shell/format.sh [hunk 8]
```diff
@@ -1,5 +1,5 @@
 #!/bin/bash
-set -Eeuo pipefail
+set -Eeu pipefail
```
Comment (fchollet) [inline shell/format.sh:2]: You can revert this

File: keras/ops/core_test.py [hunk 9]
```diff
@@ -215,6 +241,53 @@ def test_slice_update(self):
...
+        ys_op = ys_op.tolist()
+        ys_jax = ys_jax.tolist()
+
+        self.assertEqual(carry_op, carry_jax)
+
+        for i in range(len(ys_op)):
+            ys_op[i] = round(ys_op[0],2)
+            ys_jax[i] = round(ys_jax[0], 2)
+
+        self.assertListEqual(ys_op, ys_jax)
```
Comment (fchollet) [inline keras/ops/core_test.py:289]: Use `assertAllClose` for numerical checks.

Please test more input argument configurations. You could use a parameterize test for that.

**Review by fchollet — COMMENTED** [review]
Thanks for the updates!

File: keras/ops/core_test.py (hunk 9 above)
Comment (vulkomilev) [reply keras/ops/core_test.py]: New test added. I couldn't use parameterize because I can't pass arguments.I used instead for cycle.

**Comment by vulkomilev** [comment]
New commit

**Comment by fchollet** [comment]
Thanks for the update. Can you get the tests to pass?

**Comment by vulkomilev** [comment]
I think that test are passing now @fchollet

File: keras/ops/core.py [hunk 10]
```diff
@@ -168,6 +168,14 @@ def compute_output_spec(self, inputs, start_indices, updates):
         return KerasTensor(inputs.shape, dtype=inputs.dtype)
 
 
+class Scan(Operation):
+    def call(f, init, xs, length, reverse, unroll):
+        return backend.core.scan(f, init, xs, length, reverse, unroll)
+
+    def compute_output_spec(self, f, init, xs, length, reverse, unroll):
+        return KerasTensor(xs.shape, dtype=xs.dtype)
```
Comment (fchollet) [inline keras/ops/core.py:176]: This doesn't look correct -- it should match the shape of the output tensors. To start with, there are 2 tensors in this case.

File: keras/ops/core.py [hunk 11]
```diff
@@ -168,6 +168,14 @@ def compute_output_spec(self, inputs, start_indices, updates):
         return KerasTensor(inputs.shape, dtype=inputs.dtype)
 
 
+class Scan(Operation):
+    def call(f, init, xs, length, reverse, unroll):
+        return backend.core.scan(f, init, xs, length, reverse, unroll)
```
Comment (fchollet) [inline keras/ops/core.py:173]: `unroll` is not present for most backends so this would fail.

File: keras/ops/core.py [hunk 12]
```diff
@@ -204,6 +212,46 @@ def slice_update(inputs, start_indices, updates):
     return backend.core.slice_update(inputs, start_indices, updates)
 
 
+@keras_export("keras.ops.scan")
+def scan(f, init, xs, length=None, reverse=False):
```
Comment (fchollet) [inline keras/ops/core.py:216]: You can just add `unroll` everywhere for consistency, even if it's not used by some backends.

File: keras/ops/core_test.py (hunk 3 above)
Comment (fchollet) [reply keras/ops/core_test.py:244]: The test class is called `CoreOpsStaticShapeTest` and is right above this one. You can take a look at some of the test cases for the class to get an idea of how static shape inference testing works.

**Comment by gbaned** [comment]
Hi @vulkomilev Can you please check @fchollet's comments and resolve conflicts?. Thank you!

**Comment by vulkomilev** [comment]
yes I will do it today

File: keras/ops/core.py (hunk 10 above)
Comment (vulkomilev) [reply keras/ops/core.py:176]: okay I have added 2 tensors to the return statement  but I am not sure if they are correct

**Comment by vulkomilev** [comment]
@fchollet I think that I have scan test in CoreOpsStaticShapeTest

File: keras/backend/tensorflow/core.py [hunk 13]
```diff
@@ -226,6 +226,38 @@ def slice_update(inputs, start_indices, updates):
...
+            tf.cast(init, dtype=tf.double),
+            tf.zeros_like(0, dtype=tf.double),
+        )
+        xs = tf.cast(xs, dtype=tf.double)
+    else:
+        init = (tf.cast(init, dtype=tf.int64), tf.zeros_like(0, dtype=tf.int64))
+
+    carry, ys = tf.scan(f, xs, initializer=init)
+
+    # if carry[0].dtype is tf.float64:
```
Comment (fchollet) [inline keras/backend/tensorflow/core.py:251]: If the code is no longer useful, please remove it

File: keras/ops/core.py [hunk 14]
```diff
@@ -205,6 +216,46 @@ def slice_update(inputs, start_indices, updates):
...
+        the ys array will be reversed at the end.
+        unroll: Optional positive int specifying, in the underlying operation of
+         the scan primitive, how many scan iterations to unroll within a single
+         iteration of a loop(Supported only on jax backend).
+
+    Returns:
+        A scanned array and a carry element.
+    """
+    if any_symbolic_tensors((init, xs)):
+        return SliceUpdate().symbolic_call(f, init, xs, length, reverse, unroll)
```
Comment (fchollet) [inline keras/ops/core.py:255]: So this should be `Scan`. Then there should be a `class Scan` object definition, similar to the other classes in this file

**Review by fchollet — COMMENTED** [review]
Congrats on getting the tests to pass! This would still need a `Scan` class and tests for static shape inference.

**Comment by gbaned** [comment]
Hi @vulkomilev Can you please check @fchollet's comments and keep us posted ? Thank you!

**Comment by vulkomilev** [comment]
@gbaned done

**Pull request closed without merging**